raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1555,28 +1550,65 @@ public void poll() throws IOException {
     }
 
     private void handlePendingAppends(long currentTimeMs) {
-        UnwrittenAppend unwrittenAppend = unwrittenAppends.poll();
-        if (unwrittenAppend == null || unwrittenAppend.isCancelled())
-            return;
-
         if (quorum.isLeader()) {
-            if (unwrittenAppend.isTimedOut(currentTimeMs)) {
-                unwrittenAppend.fail(new TimeoutException());
-            } else {
-                LeaderState leaderState = quorum.leaderStateOrThrow();
-                int epoch = quorum.epoch();
-                LogAppendInfo info = maybeAppendAsLeader(leaderState, unwrittenAppend.records);
-                if (info != null) {
-                    OffsetAndEpoch offsetAndEpoch = new OffsetAndEpoch(info.lastOffset, epoch);
-                    unwrittenAppend.complete(offsetAndEpoch);
-                    uncommittedAppends.put(offsetAndEpoch, new AppendBatchAndTime(info.lastOffset - info.firstOffset + 1, currentTimeMs));
+            int count = 0;
+            while (!unwrittenAppends.isEmpty() && count < MAX_RECORDS_DRAINED_PER_ITERATION) {
+                final UnwrittenAppend unwrittenAppend = unwrittenAppends.poll();
+
+                if (unwrittenAppend.isCancelled())
+                    continue;
+
+                if (unwrittenAppend.isTimedOut(currentTimeMs)) {
+                    unwrittenAppend.fail(new TimeoutException());
                 } else {
-                    unwrittenAppend.fail(new InvalidRequestException("Leader refused the append"));
+                    LeaderState leaderState = quorum.leaderStateOrThrow();
+                    int epoch = quorum.epoch();
+                    LogAppendInfo info = maybeAppendAsLeader(leaderState, unwrittenAppend.records);
+                    if (info != null) {
+                        OffsetAndEpoch offsetAndEpoch = new OffsetAndEpoch(info.lastOffset, epoch);
+
+                        if (unwrittenAppend.ackMode == AckMode.LEADER) {
+                            unwrittenAppend.complete(offsetAndEpoch);
+                        } else if (unwrittenAppend.ackMode == AckMode.QUORUM) {
+                            CompletableFuture<Long> future = new CompletableFuture<>();
+                            appendPurgatory.await(future, offsetAndEpoch.offset, unwrittenAppend.requestTimeoutMs);
+
+                            future.whenComplete((completionTimeout, exception) -> {
```

Comment:
nit: I think `completionTimeout` is a timestamp? Maybe `completionTimeMs` would be clearer?